### PR TITLE
NLP: JAX Machine Translation

### PR DESCRIPTION
**Pull request by dillonroach**
Keras/TF -> Jax/Flax.nnx convert and expand of 

[https://keras.io/examples/nlp/neural_machine_translation_with_transformer/](https://keras.io/examples/nlp/neural_machine_translation_with_transformer/)

English-to-spanish encoder/decoder transformer training and inference.

### CODE REVIEW

File: docs/JAX_machine_translation.md [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# NLP: JAX Machine Translation
+
+```{code-cell} ipython3
+## adapted from https://keras.io/examples/nlp/neural_machine_translation_with_transformer/, which is itself an adaptation from https://www.manning.com/books/deep-learning-with-python-second-edition
+```
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:19]: This part should be a text cell and not the code cell

File: docs/JAX_machine_translation.md [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+import numpy as np
+
+import jax
+import jax.numpy as jnp
+import optax
+
+from flax import nnx
+
+import tiktoken
+```
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:37]: It may make sense to add a subsection with additional requirements description and installation: for example tiktoken and grain, tqdm

File: docs/JAX_machine_translation.md [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+print(f"{len(train_pairs)} training pairs")
+print(f"{len(val_pairs)} validation pairs")
+print(f"{len(test_pairs)} test pairs")
+```
+
+```{code-cell} ipython3
+tokenizer = tiktoken.get_encoding("cl100k_base")
+```
+
+We strip out puctuation to keep things simple and in line with the original tutorial - the `[` `]` are kept in so that our `[start]` and `[end]` formatting is preserved.
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:92]:
```suggestion
We strip out punctuation to keep things simple and in line with the original tutorial - the `[` `]` are kept in so that our `[start]` and `[end]` formatting is preserved.
```

File: docs/JAX_machine_translation.md [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+At this point we've extracted the data, applied formatting, and tokenized the phrases with padding. The data is kept in train/validate/test sets that each have dictionary entries, which look like the following:
+
+```{code-cell} ipython3
+## data selection example
+print(train_data[135])
+```
+
+## Define Transformer components: Encoder, Decoder, Positional Embed
+
+In many ways this is very similar to the original source, with `ops` changing to `jnp` and `keras` or `layers` becoming `nnx`. Certain module-specific arguments come and go, like the rngs attached to most things in the updated version, and decode=False in the MHA call.
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:144]: Maybe, say explicitly what MHA means

File: docs/JAX_machine_translation.md [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+        self.embed_dim = embed_dim
+        self.dense_dim = dense_dim
+        self.num_heads = num_heads
+
+        self.attention = nnx.MultiHeadAttention(num_heads=num_heads,
+                                          in_features=embed_dim,
+                                          decode=False,
+                                          rngs=rngs) 
+        self.dense_proj = nnx.Sequential(
+            *[
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:158]: Technically, you do not need `*` and `[` `]`, just provide layers with comma between them

File: docs/JAX_machine_translation.md [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+                                  in_features=embed_dim,
+                                  decode=False,
+                                  rngs=rngs)
+        self.attention_2 = nnx.MultiHeadAttention(num_heads=num_heads,
+                                  in_features=embed_dim,
+                                  decode=False,
+                                  rngs=rngs)
+
+        self.dense_proj = nnx.Sequential(
+            *[
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:218]: Same here about `*[]`

File: docs/JAX_machine_translation.md [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+                nnx.Linear(embed_dim, latent_dim, rngs=rngs),
+                nnx.relu,
+                nnx.Linear(latent_dim, embed_dim, rngs=rngs),
+            ]
+        )
+        self.layernorm_1 = nnx.LayerNorm(embed_dim, rngs=rngs) ## check inputs
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:224]: Remove the comment `## check inputs`

File: docs/JAX_machine_translation.md [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+
+    def __call__(self, inputs, encoder_outputs, mask=None):
+        causal_mask = self.get_causal_attention_mask(inputs.shape[1])
+        if mask is not None:
+            padding_mask = jnp.expand_dims(mask, axis=1).astype(jnp.int32)
+            padding_mask = jnp.minimum(padding_mask, causal_mask)
+        else:
+            padding_mask = None
+        attention_output_1 = self.attention_1(
+            inputs_q=inputs, inputs_v=inputs, inputs_k=inputs,  mask=causal_mask  ## check kwargs, specifically qkv
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:236]: How about this, still need to check kwargs?

File: docs/JAX_machine_translation.md [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+        self.num_batches = len(data) // batch_size
+
+    def __iter__(self):
+        for i in range(self.num_batches):
+            batch = self.data[i * self.batch_size: (i + 1) * self.batch_size]
+            encoder_inputs = [item["encoder_inputs"] for item in batch]
+            decoder_inputs = [item["decoder_inputs"] for item in batch]
+            target_output = [item["target_output"] for item in batch]
+            yield (jnp.array(encoder_inputs), jnp.array(decoder_inputs), jnp.array(target_output))
+```
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:276]: Let's use grain for the data loading instead of self crafted class as IMO we have to show best practices such that reader can take this code and reuse it for their own use-cases with larger datasets etc.

File: docs/JAX_machine_translation.md [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+
+Here we finally use our earlier encoder, decoder, and positional embed classes to construct the Model that we'll train and later use for inference.
+
+```{code-cell} ipython3
+class TransformerModel(nnx.Module):
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:281]: I think model's definition should be near other layers implementations

File: docs/JAX_machine_translation.md [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+
+Optax doesn't have the identical loss function that the source tutorial uses, but this softmax cross entropy works well here - you can one_hot_encode if you don't use the `_with_integer_labels` version of the loss.
+
+```{code-cell} ipython3
+def compute_loss(logits, labels):
+    loss = optax.softmax_cross_entropy_with_integer_labels(logits=logits, labels=labels)
+    return jnp.mean(loss)
+```
+
+While in the original tutorial most of the model and training details happen inside keras, me make them explicit here in our step functions, which are later used in `train_one_epoch` and `eval_model`.
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:317]:
```suggestion
While in the original tutorial most of the model and training details happen inside keras, we make them explicit here in our step functions, which are later used in `train_one_epoch` and `eval_model`.
```

File: docs/JAX_machine_translation.md [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+    labels = batch[2]
+
+    eval_metrics.update(
+        loss=loss,
+        logits=logits,
+        labels=labels,
+    )    
+```
+
+Here, nnx.MultiMetric helps us keep track of general training statistics, while we make our own dictionaries to hold historical values
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:345]:
```suggestion
Here, `nnx.MultiMetric` helps us keep track of general training statistics, while we make our own dictionaries to hold historical values
```

File: docs/JAX_machine_translation.md [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+
+```{code-cell} ipython3
+import matplotlib.pyplot as plt
+
+plt.plot(train_metrics_history["train_loss"], label="Loss value during the training")
+plt.yscale('log')
+plt.legend()
+```
+
+And eval set Loss and Accuracy - Accuracy does continue to rise, though it's hard-earned progress.
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:444]: Maybe you can also say that model is largely overfitting after the 5th epoch...

File: docs/JAX_machine_translation.md [hunk 14]
```diff
@@ -0,0 +1,502 @@
...
+test_result_pairs = []
+for _ in range(10):
+    input_sentence = random.choice(test_eng_texts)
+    translated = decode_sequence(input_sentence)
+    
+    test_result_pairs.append(f"[Input]: {input_sentence} [Translation]: {translated}") 
+```
+
+## Test Results
+For the model and the data, not too shabby - It's definitely spanish-ish. Though when 'making' friends, please don't confuse 'hacer' (to make) with 'comer' (to eat).
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:493]: "comer amigos" :)))

File: docs/JAX_machine_translation.md [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+
+```{code-cell} ipython3
+for i in test_result_pairs:
+    print(i)
+```
+
+```{code-cell} ipython3
```
Comment (vfdev-5) [inline docs/JAX_machine_translation.md:500]: We may need to remove all empty cells

**Review by vfdev-5 — COMMENTED** [review]
@dillonroach left few comments on how to make it a bit better

**Comment by dillonroach** [comment]
Alright - adapted your comments @vfdev-5 - will get the dataloader changes in as a secondary commit

**Review by jakevdp — APPROVED** [review]
Thanks - looks great!
